File: src/test/java/com/fifo/ticketing/domain/seat/service/SeatServiceTests.java
package com.fifo.ticketing.domain.seat.service;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.BDDMockito.given;

import com.fifo.ticketing.domain.performance.entity.Category;
import com.fifo.ticketing.domain.performance.entity.Grade;
import com.fifo.ticketing.domain.performance.entity.Performance;
import com.fifo.ticketing.domain.performance.entity.Place;
import com.fifo.ticketing.domain.seat.entity.Seat;
import com.fifo.ticketing.domain.seat.entity.SeatStatus;
import com.fifo.ticketing.domain.seat.repository.SeatRepository;
import com.fifo.ticketing.global.entity.File;
import com.fifo.ticketing.global.exception.AlertDetailException;
import com.fifo.ticketing.global.exception.ErrorCode;
import java.time.LocalDateTime;
import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.test.context.ActiveProfiles;

@ActiveProfiles("ci")
@ExtendWith(MockitoExtension.class)
class SeatServiceTests {

    @InjectMocks
    private SeatService seatService;

    @Mock
    private SeatRepository seatRepository;

    private Place place;
    private Performance mockPerformance;
    private Grade modckGrade;


    @BeforeEach
    void setUp() {

        place = new Place(1L, "서울특별시 서초구 서초동 1307", "강남아트홀", 100);
        File mockFile = File.builder()
            .id(1L)
            .encodedFileName("poster.jpg")
            .originalFileName("sample.jpg")
            .build();

        mockPerformance = Performance.builder()
            .id(1L)
            .title("라따뚜이")
            .description("라따뚜이는 픽사의 영화입니다.")
            .place(place)
            .startTime(LocalDateTime.of(2025, 6, 1, 19, 0))
            .endTime(LocalDateTime.of(2025, 6, 1, 21, 0))
            .category(Category.MOVIE)
            .performanceStatus(false)
            .deletedFlag(false)
            .reservationStartTime(LocalDateTime.of(2025, 5, 12, 19, 0))
            .file(mockFile)
            .build();

        modckGrade = Grade.builder()
            .id(1L)
            .grade("A")
            .place(place)
            .defaultPrice(5000)
            .seatCount(10)
            .build();


    }

    @Test
    @DisplayName("validateBookSeats - 좌석이 AVAILABLE 상태면 BOOKED 상태로 바꿔서 반환한다 ")
    void validateBookSeats_success() {

        List<Long> seatIds = List.of(1L, 2L, 3L);

        Seat mockSeat1 = new Seat(1L, mockPerformance, "A1", 5000, modckGrade,
            SeatStatus.AVAILABLE);
        Seat mockSeat2 = new Seat(2L, mockPerformance, "A1", 5000, modckGrade,
            SeatStatus.AVAILABLE);
        Seat mockSeat3 = new Seat(3L, mockPerformance, "A1", 5000, modckGrade,
            SeatStatus.AVAILABLE);

        List<Seat> mockSeats = List.of(mockSeat1, mockSeat2, mockSeat3);

        given(seatRepository.findAllByIdInWithOptimisticLock(seatIds))
            .willReturn(mockSeats);

        List<Seat> resultSeats = seatService.validateBookSeats(seatIds);

        assertEquals(mockSeat1.getId(), resultSeats.get(0).getId());
        assertEquals(mockSeat2.getId(), resultSeats.get(1).getId());
        assertEquals(mockSeat3.getId(), resultSeats.get(2).getId());

        assertEquals(resultSeats.get(0).getSeatStatus(), SeatStatus.BOOKED);
        assertEquals(resultSeats.get(1).getSeatStatus(), SeatStatus.BOOKED);
        assertEquals(resultSeats.get(2).getSeatStatus(), SeatStatus.BOOKED);

    }


    @Test
    @DisplayName("validateBookSeats - 좌석이 이미 BOOKED 상태면 AlertDetailException이 터진다")
    void validateBookSeats_Exception_success() {

        List<Long> seatIds = List.of(1L, 2L);

        Seat mockSeat1 = new Seat(1L, mockPerformance, "A1", 5000, modckGrade,
            SeatStatus.AVAILABLE);
        Seat mockSeat2 = new Seat(2L, mockPerformance, "A1", 5000, modckGrade, SeatStatus.BOOKED);

        List<Seat> mockSeats = List.of(mockSeat1, mockSeat2);

        given(seatRepository.findAllByIdInWithOptimisticLock(seatIds))
            .willReturn(mockSeats);

        AlertDetailException exception = assertThrows(AlertDetailException.class,
            () -> seatService.validateBookSeats(seatIds));

        assertTrue(exception.getMessage().contains("2번 좌석은 이미 예약되었습니다."));
        assertEquals(ErrorCode.SEAT_ALREADY_BOOKED, exception.getErrorCode());
    }
}

File: src/test/java/com/fifo/ticketing/domain/seat/service/SeatServiceTests.kt
package com.fifo.ticketing.domain.seat.service

import com.fifo.ticketing.domain.book.entity.Book
import com.fifo.ticketing.domain.book.entity.BookSeat
import com.fifo.ticketing.domain.performance.entity.Category
import com.fifo.ticketing.domain.performance.entity.Grade
import com.fifo.ticketing.domain.performance.entity.Performance
import com.fifo.ticketing.domain.performance.entity.Place
import com.fifo.ticketing.domain.seat.entity.Seat
import com.fifo.ticketing.domain.seat.entity.SeatStatus
import com.fifo.ticketing.domain.seat.repository.SeatRepository
import com.fifo.ticketing.global.entity.File
import com.fifo.ticketing.global.exception.AlertDetailException
import com.fifo.ticketing.global.exception.ErrorCode
import com.fifo.ticketing.global.exception.ErrorException
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.matchers.shouldBe
import io.mockk.*
import jakarta.persistence.EntityManager
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.time.LocalDateTime

class SeatServiceTests {

    val seatRepository = mockk<SeatRepository>()

    val entityManager = mockk<EntityManager>()

    val seatService = SeatService(
        seatRepository, entityManager
    )

    private lateinit var mockplace: Place
    private lateinit var mockFile: File
    private lateinit var mockPerformance: Performance
    private lateinit var mockGrade: Grade
    private lateinit var mockSeat: Seat

    @BeforeEach
    fun setUp() {
        mockplace = Place(1L, "서울특별시 서초구 서초동 1307", "강남아트홀", 100)

        mockFile = File(1L, "poster.jpg", "sample.jpg")

        mockPerformance = Performance(
            1L, "라따뚜이", "라따뚜이는 픽시의 영화입니다.", mockplace,
            LocalDateTime.of(2025, 6, 1, 19, 0),
            LocalDateTime.of(2025, 6, 1, 21, 0),
            Category.MOVIE,
            false,
            false,
            LocalDateTime.of(2025, 5, 12, 19, 0),
            mockFile
        )

        mockGrade = Grade(1L, mockplace, "A", 5000, 10)

        mockSeat = Seat(1L, mockPerformance, "A1", 5000, mockGrade, SeatStatus.AVAILABLE)
    }

    @Test
    fun `validateBookSeats - 좌석이 AVAILABLE 상태면 BOOKED 상태로 바꿔서 반환한다`() {
        val seatIds = listOf(1L, 2L, 3L)

        val mockSeat1 = Seat(1L, mockPerformance, "A1", 5000, mockGrade, SeatStatus.AVAILABLE)
        val mockSeat2 = Seat(2L, mockPerformance, "A2", 5000, mockGrade, SeatStatus.AVAILABLE)
        val mockSeat3 = Seat(3L, mockPerformance, "A3", 5000, mockGrade, SeatStatus.AVAILABLE)

        every { seatRepository.findAllByIdInWithOptimisticLock(seatIds) } returns listOf( mockSeat1, mockSeat2, mockSeat3 )
        every { seatRepository.flush() } just Runs


        val result = seatService.validateBookSeats(seatIds)

        3 shouldBe result.size
        SeatStatus.BOOKED shouldBe result[0].seatStatus
        SeatStatus.BOOKED shouldBe result[1].seatStatus
        SeatStatus.BOOKED shouldBe result[2].seatStatus
    }

    @Test
    fun `validateBookSeats - 좌석이 이미 BOOKED 상태면 AlertDetailException이 터진다`() {
        val seatIds = listOf(1L, 2L)

        val mockSeat1 = Seat(1L, mockPerformance, "A1", 5000, mockGrade, SeatStatus.BOOKED)
        val mockSeat2 = Seat(2L, mockPerformance, "A2", 5000, mockGrade, SeatStatus.BOOKED)

        every { seatRepository.findAllByIdInWithOptimisticLock(seatIds) } returns listOf( mockSeat1, mockSeat2 )
        every { seatRepository.flush() } just Runs

        val exception = assertThrows<AlertDetailException> {
            seatService.validateBookSeats(seatIds)
        }

        exception.errorCode shouldBe ErrorCode.SEAT_ALREADY_BOOKED

    }

    @Test
    fun `changeSeatStatus - 입력 값에 따라 좌석 상태가 변경된다`() {
        val mockBook = mockk<Book>()
        val bookSeat = BookSeat(1, mockBook, mockSeat)
        val bookSeatList = listOf(bookSeat)

        seatService.changeSeatStatus(bookSeatList, SeatStatus.OCCUPIED)

        mockSeat.seatStatus shouldBe SeatStatus.OCCUPIED
    }

    @Test
    fun `changeSeatStatus - 존재하지 않는 좌석 상태입려되면 예외가 발생한다`() {
        val mockBook = mockk<Book>()
        val bookSeat = BookSeat(1, mockBook, mockSeat)
        val bookSeatList = listOf(bookSeat)

        val exception = shouldThrow<ErrorException> {
            seatService.changeSeatStatus(
                bookSeatList,
                SeatStatus.DELETED
            )
        }

        exception.errorCode shouldBe ErrorCode.NOT_FOUND_SEAT_STATUS
    }

    @Test
    fun `getSeatsForPerformance - 공연 ID에 해당하는 좌석을 찾고 DTO로 반환한다`() {
        val performanceId = 1L

        every { seatRepository.findValidSeatsByPerformanceId(performanceId) } returns listOf(
            mockSeat
        )

        val result = seatService.getSeatsForPerformance(performanceId)

        result.size shouldBe 1
        result[0].seatId shouldBe 1
    }

    @Test
    fun `createSeats - 좌석 리스트를 EntityManager를 통해 배치 저장한다`() {
        val seatList = (1..5).map {
            Seat(it.toLong(), mockPerformance, "A$it", 5000, mockGrade, SeatStatus.AVAILABLE)
        }

        every { entityManager.persist(any<Seat>()) } just Runs
        every { entityManager.flush() } just Runs
        every { entityManager.clear() } just Runs

        seatService.createSeats(seatList)

        seatList.forEach { seat ->
            verify { entityManager.persist(seat) }
        }

        verify(exactly = 1) { entityManager.flush() }
        verify(exactly = 1) { entityManager.clear() }
    }

}